README.md
MetFragR
========

R package for MetFrag

Details on http://c-ruttkies.github.io/MetFrag/projects/metfragr/

Installation
------------

##### Installing the package locally
- checkout the MetFragR git repository
- build the package (on command line): <br>
```bash
R CMD check metfRag
R CMD build metfRag
```
- install the package (in R): <br>
```R
install.packages("metfRag",repos=NULL,type="source")
library(metfRag)
```

##### Installing the package directly from github (in R)
```R
library(devtools)
install_github("c-ruttkies/MetFragR/metfRag")
```
- if you get an error like "ERROR: loading failed for 'i386'" on windows
due to rJava issues, disable multiarch builds:
```R
install_github("c-ruttkies/MetFragR/metfRag", args="--no-multiarch")
```


##### Example (Run MetFrag)
```R
settingsObject<-list()
settingsObject[["DatabaseSearchRelativeMassDeviation"]]<-5.0
settingsObject[["FragmentPeakMatchAbsoluteMassDeviation"]]<-0.001
settingsObject[["FragmentPeakMatchRelativeMassDeviation"]]<-5.0
settingsObject[["MetFragDatabaseType"]]<-"PubChem"
settingsObject[["NeutralPrecursorMass"]]<-253.966126
settingsObject[["PeakList"]]<-matrix(c(
90.97445, 681,
106.94476, 274,
110.02750, 110,
115.98965, 95,
117.98540, 384,
124.93547, 613,
124.99015, 146,
125.99793, 207,
133.95592, 777,
143.98846, 478,
144.99625, 352,
146.00410, 999,
151.94641, 962,
160.96668, 387,
163.00682, 782,
172.99055, 17,
178.95724, 678,
178.97725, 391,
180.97293, 999,
196.96778, 720,
208.96780, 999,
236.96245, 999,
254.97312, 999), ncol=2, byrow=TRUE)
settingsObject[["NeutralPrecursorMolecularFormula"]]<-"C7H5Cl2FN2O3"
settingsObject[["PrecursorCompoundIDs"]]<-c("50465", "57010914", "56974741", "88419651", "23354334")
#
# pre and post-processing filter
#
# define filters to filter unconnected compounds (e.g. salts)
settingsObject[["MetFragPreProcessingCandidateFilter"]]<-c("UnconnectedCompoundFilter","IsotopeFilter")
settingsObject[["MetFragPostProcessingCandidateFilter"]]<-c("InChIKeyFilter")
#run MetFrag
scored.candidates<-run.metfrag(settingsObject)
#scored.candidates is a data.frame with scores and candidate properties
```



metfRag/NEWS
CHANGES IN VERSION 2.4.2
------------------------
 o added more properties to candidate list when running run.metfrag
 o java library updated to newest version 2.4.2

CHANGES IN VERSION 2.0
------------------------
 o Functions in metfRag.R redesigned due to various problems
 o database.R removed as database search is enabled directly within the java library
 o java library updated to newest version 2.3

CHANGES IN VERSION 1.0.4
------------------------
 o Add parameterized image resolution to plotMol()
 o Modify rectangle plotting in dendrogram to support k=N "clusters"

CHANGES IN VERSION 1.0.3
------------------------
 o Added vignette, visualisation of cluster representatives

CHANGES IN VERSION 1.0.2
------------------------
 o Allow parameters to be passed down to dendromat()

CHANGES IN VERSION 1.0.1
------------------------
 o Added plotCluster() function

CHANGES IN VERSION 1.0.0
------------------------
 o Initial release


metfRag/R/tools.R
#' MetFrag result tools
#' 
#' Small tools to extract information from MetFrag or MetFusion results
#' 
#' @aliases getScores
#' @param mols A list of rCDK \code{molecules} with MetFrag or MetFusion results
#' @param scoreprop The name of the property of the molecules where the score is kept
#' @export

getScores <- function(mols, scoreprop="Score"){
  return (sapply(mols, function(x) as.numeric(get.property(x, scoreprop))))
}

#' MetFrag result tools
#' 
#' Small tools to extract information from MetFrag or MetFusion results
#' 
#' @aliases getPeaksExplained 
#' @param mols A list of rCDK \code{molecules} with MetFrag or MetFusion results
#' @export
getPeaksExplained <- function(mols){
  return (sapply(mols, function(x) as.numeric(get.property(x, "NoExplPeaks"))))
}

#' MetFrag result tools
#' 
#' Small tools to extract information from MetFrag or MetFusion results
#' 
#' @aliases getDatabaseIDs
#' @param mols A list of rCDK \code{molecules} with MetFrag or MetFusion results
#' @param idprop The name of the property of the molecules where the database ID is kept
#' @export
getDatabaseIDs <- function(mols, idprop="Identifier") {
  return (sapply(1:length(mols), function(i) {
    id <- get.property(mols[[i]], idprop)
    if (is.na(id) || id == 'NA')
      return (i)
    else
      return (get.property(mols[[i]], idprop))
  }))
}

#' Plotting a molecule
#' 
#' rCDK provides either functions to plot into a Java window,
#' or into a rasterImage object. plotMol
#' is a high-level plotting function to plot into an R device.
#' 
#' @aliases plotMol
#' @param mol An rCDK \code{mol} object
#' @param smiles A SMILES string, ignored if mol is supplied
#' @param width,height Number of pixels for molecule image in x,y
#' @param watermark A string written across the image as watermark
#' @export
plotMol <- function(mol=NULL, smiles=NULL, width = 200, height = 200, watermark=NULL) {
  if (missing(mol)) {
    mol <- parse.smiles(smiles)[[1]]
  }
  #  img <- view.image.2d(mol, width = 1024, height = 1024)
  img <- view.image.2d(mol, width = width, height = height)
  plot.new()
  plot.window(xlim=c(0,1), ylim=c(0,1))
  if (!is.null(watermark)) {
    text(x=0.5, y=0.5, label=watermark, 
         cex=0.55/strwidth(watermark), 
         col=rgb(0.75, 1,0.75))
  }
  rasterImage(alpha2image(img), 0,0, 1,1)
}

#' Image manipulation
#' 
#' This function adds an alpha channel to an image
#' 
#' @aliases alpha2image
#' @param img A rasterImage
#' @param threshold An intensity threshold, all pixels brighter than this 
#' will be set to transparent
#' @export
alpha2image <- function(img, threshold=0.5) {
  a2 <- as.array(img)
  th <- 0.5 ## Threshhold
  alpha <- a2[,,1]<th | a2[,,2]<th | a2[,,3]<th
  
  d <- dim(a2)
  d[3] <- 4
  a3 <- array(dim=d)
  
  a3[,,1:3] <- a2
  a3[,,4] <- alpha
  
  img <- as.raster(a3)  
}
  

#' Chemical Clustering
#' 
#' Calculate the Maximum Common Substructure for a set of molecules.
#' This is a star-alignment, iterating through all structures
#' and creating the consensus structure as Maximum Common Substructure 
#' between the consensus and the next molecule in the list.
#' 
#' @aliases getMCSS
#' @usage getMCSS(mols)
#' @param mols The a list of rCDK \code{mols}
#' @author Steffen Neumann (\email{sneumann@ipb-halle.de})
#' @examples 
#'        library(rcdk)
#'        smiles <- c('CCC', 'CCN', 'CCN(C)(C)',
#'                    'c1ccccc1Cc1ccccc1',
#'                    'C1CCC1CC(CN(C)(C))CC(=O)CC')
#'        mols <- parse.smiles(smiles)
#'        dummy <- getMCSS(mols)
#' 
#' @export

getMCSS <- function(mols) {
  l <- length(mols)
  if (l<1)
    stop("No MCSS for less than one molecule!")
  
  if (l==1)
    return (mols[[1]])
  
  consensus <- mols[[1]]  
  for (i in 2:l) {
    consensus <- get.mcs(consensus, mols[[i]])
  }
  return (consensus)  
}

#' Retrieve PubChem Trivial Names
#' 
#' Fetches trivial names for a processed PubChem candidate list retrieved by \code{run.metfrag} command.
#' A new candidate list is returned with an additional column including the trivial names
#' from PubChem. 
#'
#' @aliases add.trivialname.pubchem
#' @usage add.trivialname.pubchem(result.list.metfrag)
#' @param processed candidate list retrieved by \code{run.metfrag} and PubChem
#' @author Christoph Ruttkies (\email{cruttkie@ipb-halle.de})
#' @examples 
#'        obj <- create.settings.sample()
#'        result.list.metfrag <- run.metfrag(smiles)
#'        result.list.metfrag <- add.trivialname.pubchem(result.list.metfrag)
#' 
#' @export
add.trivialname.pubchem <- function(result.list.metfrag) {
  if(!require(rjson)) {
    print("Please install package: rjson")
    return(result.list.metfrag)
  }
  ids <- as.character(result.list.metfrag[,"Identifier"])
  names <- rep("NA", dim(result.list.metfrag)[1])
  names(names) <- ids
  id_string <- ""
  packages <- ceiling(length(ids) / 200)
  for(i in 1:length(ids)) {
    if(id_string != "") {
      id_string <- paste(id_string, ids[i], sep=",")
    } else {
      id_string <- paste(id_string, ids[i], sep="")
    }
    if(i %% 200 == 0 || i == length(ids)) {
      link <- paste("http://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/", id_string, "/synonyms/JSON", sep="")
      document <- fromJSON(paste(readLines(link), collapse=""))
      id_string <- ""
      sapply(document[[1]][[1]], function(element) {
        if(!is.null(element$Synonym[1])) {names[as.character(element$CID)] <<- element$Synonym[1]}
      })
      cat(paste(ceiling(i / 200),"/" , packages, " ", sep=""), append=T)
    }
  }
  cat("", sep="\n")
  if("TrivialName" %in% colnames(result.list.metfrag)) {
    named.result.list.metfrag <- result.list.metfrag
    named.result.list.metfrag[,"TrivialName"] <- names
  } else {
    named.result.list.metfrag <- cbind(result.list.metfrag, names)
    colnames(named.result.list.metfrag)[length(named.result.list.metfrag)] <- "TrivialName"
  }
  return(named.result.list.metfrag)
}



metfRag/R/common.R
common.lib.findMolecule <- function(mol, property, seek.value, split=NULL)
{
  mol.value <- get.property(mol, property);
  #stuff <- names(get.properties(mol));
  #stop(stuff)
  
  if (is.null(split) == TRUE && mol.value == seek.value)
  {
    return(TRUE);
  }
  else if (missing(split) == FALSE && 
           is.null(split$sep) == FALSE && 
           is.null(split$pos) == FALSE)
  {
      mol.value <- strsplit(mol.value,split=split$sep)[[1]][split$pos];
      
      if (mol.value == seek.value)
      { return(TRUE); }
  }
  
  return(FALSE);
}

common.lib.lookup <- function(needle, haystack, link, split=NULL)
{
  if (missing(needle) == TRUE || 
      missing(haystack) == TRUE || 
      length(haystack) == 0)
  { return(FALSE); }
  
  occur <- c();
  
  if (is.null(split) == FALSE && 
      is.null(split$sep) == FALSE && 
      is.null(split$pos) == FALSE)
  {
    mol.value.needle <- strsplit(get.property(needle, link),
                                 split=split$sep)[[1]][split$pos];    
    
    for (i in (1:length(haystack)))
    {
      mol.value.hay <- strsplit(get.property(haystack[[i]], link),
                                split=split$sep)[[1]][split$pos];      
      
      if (mol.value.needle == mol.value.hay)
      { occur <- c(occur, i); }
    }
  }
  else
  {
    for (i in (1:length(haystack)))
    {
      if (get.property(needle, link) == get.property(haystack[[i]], link))
      { occur <- c(occur, i); }
    }
  }
  
  return (occur);
}

common.lib.lookupFirst <- function(needle, haystack, link, split=NULL)
{
  if (missing(needle) == TRUE || 
      missing(haystack) == TRUE || 
      length(haystack) == 0)
  { return(FALSE); }

  if (is.null(split) == FALSE && 
        is.null(split$sep) == FALSE && 
        is.null(split$pos) == FALSE)
  {
    mol.value.needle <- strsplit(get.property(needle, link),
                                 split=split$sep)[[1]][split$pos];    
    
    for (i in (1:length(haystack)))
    {
      mol.value.hay <- strsplit(get.property(haystack[[i]], link),
                                split=split$sep)[[1]][split$pos];      
      
      if (mol.value.needle == mol.value.hay)
      { return(TRUE); }
    }
  }
  else
  {
    for (i in (1:length(haystack)))
    {
      if ((get.property(needle, link) == get.property(haystack[[i]], link)))
      { return(TRUE); }
    }
  }
  
  return(FALSE);
}

lib.CommonSet <- function(set)
{
  set.result <- c();
  
  for (i in 1:length(set))
  { set.result <<- c(set.result, names(set[[i]])); }

  set.result <- intersect(set.result, set.result);

  return(set.result);
}

comm.lib.showLinkOptions <- function(set.a, set.b)
{
  if (missing(set.a) == TRUE || missing(set.b) == TRUE)
  { return(FALSE); }
  
  set.a.prop <- sapply(set.a, get.properties);
  set.b.prop <- sapply(set.b, get.properties);
  
  set.inter <- intersect(
    lib.CommonSet(set.a.prop),
    lib.CommonSet(set.b.prop)
  );
    
  return(set.inter);
}

comm.lib.showNumberOptions <- function(set)
{
  if (missing(set) == TRUE)
  { return(FALSE); }
  
  keys <- comm.lib.showLinkOptions(set,set);
  name <- c();
  
  for (i in (1:length(keys)))
  {
    if (is.numeric(get.property(set[[1]],keys[i])))
    { name <- c(name, keys[i])}
    else
    {
      x <- as.numeric(get.property(set[[1]],keys[i]));
      
      if (is.na(x) == FALSE)
      {name <- c(name, keys[i]);}
    }
  }
  
  return(name);
}


metfRag/R/metfRag.R
#' @import rcdk
#' @import rJava
NULL

.packageName <- "metfRag"

require(rJava, quietly=TRUE)

.onLoad<-function(libname, pkgname) {
	jar.metfrag <- paste(libname, pkgname, "java", "MetFragR-2.4.2-jar-with-dependencies.jar", sep=.Platform$file.sep)	
	.jinit(classpath=c(jar.metfrag))
}

#' Create a settings object used by run.metfrag function
#' 
#' Returns a list with defined settings.
#' 
#' @author Christoph Ruttkies (\email{cruttkie@ipb-halle.de})
#' @export
create.settings.sample<-function() {
  settingsObject<-list()
  settingsObject[["DatabaseSearchRelativeMassDeviation"]]<-5.0
  settingsObject[["FragmentPeakMatchAbsoluteMassDeviation"]]<-0.001
  settingsObject[["FragmentPeakMatchRelativeMassDeviation"]]<-5.0
  settingsObject[["MetFragDatabaseType"]]<-"PubChem"
  settingsObject[["NeutralPrecursorMass"]]<-253.966126
  settingsObject[["NeutralPrecursorMolecularFormula"]]<-"C7H5Cl2FN2O3"
  settingsObject[["PrecursorCompoundIDs"]]<-c("50465", "57010914", "56974741", "88419651", "23354334")
  settingsObject[["MetFragPreProcessingCandidateFilter"]]<-c("UnconnectedCompoundFilter","IsotopeFilter")
  settingsObject[["MetFragPostProcessingCandidateFilter"]]<-c("InChIKeyFilter")
  settingsObject[["PeakList"]]<-matrix(c(	
  90.97445, 681,	
  106.94476, 274,	
  110.02750, 110,	
  115.98965, 95,	
  117.98540, 384,	
  124.93547, 613,	
  124.99015, 146,	
  125.99793, 207,	
  133.95592, 777,	
  143.98846, 478,	
  144.99625, 352,	
  146.00410, 999,	
  151.94641, 962,	
  160.96668, 387,	
  163.00682, 782,	
  172.99055, 17,	
  178.95724, 678,	
  178.97725, 391,	
  180.97293, 999,	
  196.96778, 720,	
  208.96780, 999,	
  236.96245, 999,	
  254.97312, 999), ncol=2, byrow=TRUE)
  return(settingsObject)
}

#' Run MetFrag by specifying all parameters given with the settings object as list
#' 
#' The function uses defined settings like database, fragmentation and scoring
#' parameters to initialise and run the MetFrag process. The FragmenterScore 
#' calculation is based on the match of in silico generated fragments of the
#' candidate molecules to the given tandem mass spectrum. A data.frame of candidates
#' molecules is returned sorted by their final score.
#' 
#' 
#' @param list of settings 
#' @author Christoph Ruttkies (\email{cruttkie@ipb-halle.de})
#' @export
run.metfrag<-function(settingsObject) {

	if(missing(settingsObject)) stop("Error: Settings object is missing!")
	
	if(class(settingsObject) != "list") stop("Error: Settings object must be of type list!")
	if(is.null(names(settingsObject))) stop("Error: Settings object does not contain valid names!")
	if(length(settingsObject) == 0) stop("Error: Settings object does not contain valid values!")
      
	javaSettings=.jnew("de/ipbhalle/metfraglib/settings/MetFragGlobalSettings")
	
	getDatatype<-function(name, value) {
	  vector = FALSE;
	  if(class(value) != "character" & length(value) > 1) {vector = TRUE}
	  if(name == "NeutralPrecursorMass") {return("double")}
	  else if(name == "KeggProxyPort") {return("integer")}
	  else if(name == "MoNAProxyPort") {return("integer")}
	  else if(name == "MetaCycProxyPort") {return("integer")}
	  else if(name == "PubChemProxyPort") {return("integer")}
	  else if(name == "NeutralPrecursorMass") {return("double")}
	  else if(name == "DatabaseSearchRelativeMassDeviation") {return("double")}
	  else if(name == "FragmentPeakMatchAbsoluteMassDeviation") {return("double")}
	  else if(name == "FragmentPeakMatchRelativeMassDeviation") {return("double")}
	  else if(name == "MaximumTreeDepth") {return("integer")}
	  else if(name == "PrecursorIonMode") {return("integer")}
	  else if(name == "IonizedPrecursorMass") {return("double")}
	  else if(name == "NumberThreads") {return("byte")}
	  else if(name == "ExperimentalRetentionTimeValue") {return("double")}
	  else if(name == "MinimumAbsolutePeakIntensity") {return("double")}
	  else if(name == "SmartsSubstructureExclusionScoreSmartsList") {return("array")}
	  else if(name == "SmartsSubstructureInclusionScoreSmartsList") {return("array")}
	  else if(name == "ScoreSmartsInclusionList") {return("array")}
	  else if(name == "ScoreSmartsExclusionList") {return("array")}
	  else if(name == "FilterSmartsInclusionList") {return("array")}
	  else if(name == "FilterSmartsExclusionList") {return("array")}
	  else if(name == "FilterSuspectLists") {return("array")}
	  else if(name == "ScoreSuspectLists") {return("array")}
	  else if(name == "FilterExcludedElements") {return("array")}
	  else if(name == "FilterIncludedElements") {return("array")}
	  else if(name == "CombinedReferenceScoreValues") {return("array")}
	  else if(name == "MetFragScoreWeights") {return("array_double")}
	  else if(name == "MetFragPreProcessingCandidateFilter") {return("array")}
	  else if(name == "MetFragPostProcessingCandidateFilter") {return("array")}
	  else if(name == "PrecursorCompoundIDs") {return("array")}
	  else if(name == "MetFragScoreTypes") {return("array_double")}
	  else if(name == "MetFragCandidateWriter") {return("array")}
	  else if(vector) {
	    if(class(value) == "numeric" && value == round(value)) {return("array")}
	    else if(class(value) == "numeric" && value != round(value)) {return("array_double")}
	    else if(class(value) == "character") {return("array")}
	    else if(class(value) == "logical") {return("array")}
	    else {return("unknown")}
	  }
	  else if(!vector) {
	    if(class(value) == "numeric" && value == round(value)) {return("integer")}
	    else if(class(value) == "numeric" && value != round(value)) {return("double")}
	    else if(class(value) == "character") {return("string")}
	    else if(class(value) == "logical") {return("boolean")}
	    else {return("unknown")}
	  }

	}

	#write all properties into Java settings object
	sapply(names(settingsObject), function(name) {
	  #check the peak list
	  if(name == "PeakList") {
	    if(is.null(settingsObject[[name]])) {
	      .jcall(javaSettings, "V", 'set', "PeakListString", .jnew("java.lang.String", as.character(paste(settingsObject[[name]][1], settingsObject[[name]][1], sep=" "))))
	    }
	    else {
	      .jcall(javaSettings, "V", 'set', "PeakListString", paste(apply(settingsObject[[name]], 1, function(x) paste(x[1], x[2])), collapse="\n"))
	    }
	    .jcall(javaSettings, "V", 'set', "MetFragPeakListReader", .jnew("java.lang.String", as.character("de.ipbhalle.metfraglib.peaklistreader.FilteredStringTandemMassPeakListReader")))
	  }
	  #check further settings
	  else {
	    #in case it a single value
	    if(getDatatype(name, settingsObject[[name]]) == "integer") {
	      .jcall(javaSettings, "V", 'set', name, .jnew("java.lang.Integer", as.integer(settingsObject[[name]])))
	    }
	    else if(getDatatype(name, settingsObject[[name]]) == "double") {
	      .jcall(javaSettings, "V", 'set', name, .jnew("java.lang.Double", as.double(settingsObject[[name]])))
	    }
	    else if(getDatatype(name, settingsObject[[name]]) == "string") {
	      .jcall(javaSettings, "V", 'set', name, .jnew("java.lang.String", as.character(settingsObject[[name]])))
	    }
	    else if(getDatatype(name, settingsObject[[name]]) == "boolean") {
	      .jcall(javaSettings, "V", 'set', name, .jnew("java.lang.Boolean", as.logical(settingsObject[[name]])))
	    }
	    else if(getDatatype(name, settingsObject[[name]]) == "byte") {
	      .jcall(javaSettings, "V", 'set', name, .jnew("java.lang.Byte", .jbyte(settingsObject[[name]])))
	    }
	    #vectors
	    else if(getDatatype(name, settingsObject[[name]]) == "array") {
	      .jcall(javaSettings, "V", 'set', name, .jarray(settingsObject[[name]]))
	    }
	    else if(getDatatype(name, settingsObject[[name]]) == "array_double") {
	      .jcall(javaSettings, "V", 'set', name, .jarray(settingsObject[[name]], "[java.lang.Double"))
	    }
	    else {
	      print(paste("Unknown type of parameter", name, "(", class(settingsObject[[name]]), ")"))
	    }
	  }
	  if(name == "PeakListString") {
	    .jcall(javaSettings, "V", 'set', "MetFragPeakListReader", .jnew("[java.lang.String", as.character("de.ipbhalle.metfraglib.peaklistreader.FilteredStringTandemMassPeakListReader")))
	  }
	})

	temp_dir=tempdir()
	J("java.lang.System")$setProperty( "java.io.tmpdir", temp_dir )

	obj=.jnew("de/ipbhalle/metfrag/r/MetfRag")
	candidateList=.jcall(obj, "Lde/ipbhalle/metfraglib/list/CandidateList;", "runMetFrag", javaSettings)
	candidateList=.jcast(candidateList, "de/ipbhalle/metfraglib/list/ScoredCandidateList")
	
	#remove axis temp dirs in case they were generated
	axis_files<-dir(temp_dir, full.names=T)[grep("^axis2", dir(temp_dir))]
	axis_dirs<-axis_files[which(file.info(axis_files)[,"isdir"])]
	files_to_remove<-unlist(lapply(1:length(axis_dirs), function(x) dir(axis_dirs[x], full.names=T)))
	files_to_remove<-files_to_remove[grep("jar$", files_to_remove)]
	unlink(files_to_remove)

	numberCandidates<-.jcall(candidateList, "I", "getNumberElements")
	
	if(numberCandidates == 0) return(data.frame())
	
	numberPeaksUsed<-.jcall(candidateList, "I", "getNumberPeaksUsed")
	propertyNames<-c()
	datatypes<-list()
	
	if(numberCandidates >= 1) {
	  candidate <- .jcall(candidateList, "Lde/ipbhalle/metfraglib/interfaces/ICandidate;", "getElement", as.integer(0))
	  propertyNames <- .jcall(candidate, "[S", "getPropertyNames")
	  sapply(1:length(propertyNames), function(propertyIndex) {
	    datatypes[[propertyNames[propertyIndex]]]<<-.jcall(candidate, "Ljava/lang/Object;", "getProperty", propertyNames[propertyIndex])$getClass()$getName()
	  })
	}

	candidateProperties<-list()
	datatypes[["NoExplPeaks"]] <- "java.lang.Integer"
	candidateProperties[["NoExplPeaks"]] <- vector(mode = "numeric", length = 0)	

	sapply(1:length(propertyNames), function(propertyIndex) {
	  candidateProperties[[propertyNames[propertyIndex]]] <<- vector(mode = "character", length = 0)
	})
	sapply(1:numberCandidates, function(candidateIndex) {
	  candidate <- .jcall(candidateList, "Lde/ipbhalle/metfraglib/interfaces/ICandidate;", "getElement", as.integer(candidateIndex - 1))
	  sapply(1:length(propertyNames), function(propertyIndex) {
	    value <- .jcall(candidate, "Ljava/lang/Object;", "getProperty", propertyNames[propertyIndex])$toString()
	    candidateProperties[[propertyNames[propertyIndex]]] <<- c(candidateProperties[[propertyNames[propertyIndex]]], value)
	  })
	  candidateProperties[["NoExplPeaks"]] <<- c(candidateProperties[["NoExplPeaks"]], .jcall(candidate, "I", "getNumberPeaksExplained"))
	})
	
	sapply(1:length(propertyNames), function(propertyIndex) {
	  datatype<-datatypes[[propertyNames[propertyIndex]]]
	  if(datatype == "java.lang.Double" || datatype == "java.lang.Byte" || datatype == "java.lang.Integer" || datatype == "java.lang.Float") {
	    suppressWarnings(candidateProperties[[propertyNames[propertyIndex]]]<<-as.numeric(candidateProperties[[propertyNames[propertyIndex]]]))
	  }
	})

	datatypes[["NumberPeaksUsed"]]<-"java.lang.Integer"
	candidateProperties[["NumberPeaksUsed"]] <- rep(numberPeaksUsed, numberCandidates)

	return(as.data.frame(candidateProperties))
}


#' Calculate MetFrag scores for molecules and a given tandem mass spectrum
#' 
#' The function calculates scores for molecules given in a SD file. The score
#' calculation is based on the match of in silico generated fragments of the
#' candidate molecules to the given tandem mass spectrum. A list of candidate
#' molecules is returned sorted by their MetFrag score.
#' 
#' 
#' @param sdf The name of the SD file containing candidate molecules
#' @param mzs A \code{vector} mass to charge ratio values
#' @param ints A \code{vector} of intensity values
#' @param exact.mass The neutral exact mass of the precursor molecule
#' @param number.threads Number threads for parallel execution (max. 8)
#' @param mz.abs Absolute mass deviation (Da) allowed to match theoretical
#' fragment massed to the given mz values
#' @param mz.ppm Relative mass deviation (ppm) allowed to match theoretical
#' fragment massed to the given mz values
#' @param search.ppm The relative mass deviation (ppm) from \code{exact.mass}
#' for candidate molecules in \code{sdf}
#' @param pos.charge If \code{TRUE} the given tandem mass spectrum is assumed
#' to be measured in positive mode
#' @param mode Type of the measured molecule: -1 -> [M-H], 0 -> [M], 1 -> [M+H]
#' @param tree.depth Maximal tree depth of MetFrag to generate fragments
#' @author Christoph Ruttkies (\email{cruttkie@ipb-halle.de})
#' @export score.molecules.from.sdf
score.molecules.from.sdf<-function(sdf, mzs, ints, exact.mass, number.threads=1, mz.abs=0.01, mz.ppm=10, search.ppm=10, pos.charge=TRUE, mode=1, tree.depth=2, score.names=c("FragmenterScore"), scoreWeights=c(1.0)) {
	
	if(missing(sdf)) stop("Error: SDF is missing!")
	if(missing(mzs)) stop("Error: Vector of mass to charge ratios is missing!")
	if(missing(ints)) stop("Error: Vector of intensities is missing!")
	if(missing(exact.mass)) stop("Error: Neutral exact mass of parent molecule is missing!")

	if(file.access(sdf, 4) != 0) stop("Error: Cannot access SDF! Either it does not exist or is not readable.");
	if(!is.vector(mzs) || !is.numeric(mzs) || any(mzs < 0)) stop("Error: The argument mzs must be a vector of numerals!")
	if(!is.vector(ints) || !is.numeric(ints) || any(ints < 0)) stop("Error: The argument ints must be a vector of positive numerals!")
	if(exact.mass < 0) stop("Error: The argument exact.mass must be a positive numeral!")	
	if(number.threads < 0 || number.threads > 8) stop("Error: The argument number.threads must get values from 0 till 8!")
	if(mz.abs < 0) stop("Error: The argument mz.abs must be positive!")	
	if(mz.ppm < 0) stop("Error: The argument mz.ppm must be positive!")
	if(search.ppm < 0) stop("Error: The argument search.ppm must be positive!")
	if(!is.logical(pos.charge)) stop("Error: The argument pos.charge must be a logical value!")
	if(!is.element(mode, c(-1,0,1))) stop("Error: The argument mode must be a value of -1, 0 or 1")
	if(tree.depth < 1 || tree.depth > 5) stop("Error: The argument tree.depth must get values from 1 till 5!")
	
	obj=.jnew("de/ipbhalle/metfrag/r/MetfRag")
	mols<-.jcall(obj, '[Lorg/openscience/cdk/interfaces/IAtomContainer;', 
               'scoreMoleculesAgainstSpectrum', sdf, 
               .jarray(as.double(mzs),"[D"), 
               .jarray(as.double(ints),"[D"), 
               as.double(exact.mass), as.integer(number.threads), as.double(mz.abs), as.double(mz.ppm), 
               as.double(search.ppm), as.logical(pos.charge), as.integer(mode), as.integer(tree.depth), .jarray(score.names,"[S"), .jarray(as.double(scoreWeights),"[D"))
	obj = .jnull()
	
  if(length(mols) == 0) 
    cat("No results generated!","\n")
	
  return(mols)
}

#' Calculate MetFrag scores for molecules and a given tandem mass spectrum
#' 
#' The function calculates scores for molecules given by a vector of rcdk atomcontainers. 
#' The score calculation is based on the match of in silico generated fragments 
#' of the candidate molecules to the given tandem mass spectrum. A list of candidate
#' molecules is returned sorted by their MetFrag score.
#' 
#' 
#' @param molecules candidate molecules as vector of rcdk atomcontainers
#' @param mzs A \code{vector} mass to charge ratio values
#' @param ints A \code{vector} of intensity values
#' @param exact.mass The neutral exact mass of the precursor molecule
#' @param number.threads Number threads for parallel execution (max. 8)
#' @param mz.abs Absolute mass deviation (Da) allowed to match theoretical
#' fragment massed to the given mz values
#' @param mz.ppm Relative mass deviation (ppm) allowed to match theoretical
#' fragment massed to the given mz values
#' @param pos.charge If \code{TRUE} the given tandem mass spectrum is assumed
#' to be measured in positive mode
#' @param mode Type of the measured molecule: -1 -> [M-H], 0 -> [M], 1 -> [M+H]
#' @param tree.depth Maximal tree depth of MetFrag to generate fragments
#' @author Christoph Ruttkies (\email{cruttkie@ipb-halle.de})
#' @export score.molecules.from.container
score.molecules.from.container<-function(molecules, mzs, ints, exact.mass, 
                                         number.threads=1, mz.abs=0.01, 
                                         mz.ppm=10, pos.charge=TRUE, 
                                         mode=1, tree.depth=2, score.names=c("FragmenterScore"), scoreWeights=c(1.0)) 
{
  if(missing(molecules)) stop("Error: Molecules are missing!")
  if(missing(mzs)) stop("Error: Vector of mass to charge ratios is missing!")
  if(missing(ints)) stop("Error: Vector of intensities is missing!")
  if(missing(exact.mass)) stop("Error: Neutral exact mass of parent molecule is missing!")
  
  if(!is.vector(mzs) || !is.numeric(mzs) || any(mzs < 0)) stop("Error: The argument mzs must be a vector of numerals!")
  if(!is.vector(ints) || !is.numeric(ints) || any(ints < 0)) stop("Error: The argument ints must be a vector of positive numerals!")
  if(exact.mass < 0) stop("Error: The argument exact.mass must be a positive numeral!")	
  if(number.threads < 0 || number.threads > 8) stop("Error: The argument number.threads must get values from 0 till 8!")
  if(mz.abs < 0) stop("Error: The argument mz.abs must be positive!")	
  if(mz.ppm < 0) stop("Error: The argument mz.ppm must be positive!")
  if(!is.logical(pos.charge)) stop("Error: The argument pos.charge must be a logical value!")
  if(!is.element(mode, c(-1,0,1))) stop("Error: The argument mode must be a value of -1, 0 or 1")
  if(tree.depth < 1 || tree.depth > 5) stop("Error: The argument tree.depth must get values from 1 till 5!")
  
  obj = .jnew("de/ipbhalle/metfrag/r/MetfRag")
  mols<- .jcall(obj, '[Lorg/openscience/cdk/interfaces/IAtomContainer;',
               'scoreMoleculesAgainstSpectrum', 
               .jarray(molecules, contents.class = "org/openscience/cdk/interfaces/IAtomContainer"),
               .jarray(as.double(mzs),"[D"), 
               .jarray(as.double(ints),"[D"), 
               exact.mass, as.integer(number.threads), as.double(mz.abs), as.double(mz.ppm), 
               as.logical(pos.charge), as.integer(mode), as.integer(tree.depth), .jarray(score.names,"[S"), .jarray(as.double(scoreWeights),"[D"))
  obj = .jnull()
  
  if(length(mols) == 0) 
    cat("No results generated!","\n")
  
  return(mols)
}
